Ignore conversion errors when building StructuredData

cjohnson78 opened the pull request:
Added the ability to ignore conversion errors when building StructuredData by setting structuredData.ignoreConversionErrors=true

cjohnson78 committed:
Added the ability to ignore conversion errors when building StructuredData by setting structuredData.ignoreConversionErrors=true

Sdwz commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/StructuredData.java:
```diff
@@ -354,16 +359,21 @@ private NamedProperty getProperty(String propertyName, Collection<Object> values
+              propertyName,
+              nonNullValues
+                  .stream()
+                  .map(v -> valueConverter.convert(v))
```
valueConverter::convert while you're at it :D

cjohnson78 replied:
I do not understand, what do you mean by valueConverter::convert ?

TanmayVartak replied:
v-> valueConverter.convert(v) can also be written as valueConverter::convert

I am fine with existing pattern as well :)

cjohnson78 replied:
Ok, thanks for the clarification.

Sdwz commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/StructuredData.java:
```diff
@@ -354,16 +359,21 @@ private NamedProperty getProperty(String propertyName, Collection<Object> values
+                  .collect(Collectors.toList()));
+        }
+      } catch (IllegalArgumentException e) {
+        if(ignoreConversionErrors) return null;
```
I would definitely add a log here. Having to check for every single result if a property is there without being warned this happen will be a pain.

cjohnson78 replied:
```suggestion
        if(ignoreConversionErrors) {
          logger.log(Level.FINEST, "Ignoring conversion error: {0}", e.getMessage());
          return null;
        }
```

Sdwz commented:
More generally, this should be doing the same for both repeatable and non repeatable data.

cjohnson78 reviewed (commented):
The change I made will ignore errors thrown for single and repeatable data. Do you propose a different way of handling repeating values? I feel like in general, people that use this settings have a goal of preventing the connector from stopping because of one bad value (regardless of single vs. multi.). We know there is bad data sometimes and we just need to push past it in some cases. In other cases, we would leave this disabled and we would work through the errors until we have added adequate date formats, for example.

cjohnson78 commented:
Is there any further feedback at this point?

TanmayVartak commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/StructuredData.java:
```diff
@@ -129,6 +131,7 @@
 
   private static final AtomicBoolean initialized = new AtomicBoolean();
   private static final List<DateTimeParser> dateTimeParsers = new ArrayList<>();
+  private static boolean ignoreConversionErrors = false;
```
make this AtomicBoolean

Sdwz replied:
Why would it be required here ? The method where the Boolean is set is synchronized and protected by the configuration initialized atomic boolean.

cjohnson78 replied:
I went ahead and made it AtomicBoolean, just in case it is set outside of the init in the future

TanmayVartak commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/StructuredData.java:
```diff
@@ -354,16 +359,21 @@ private NamedProperty getProperty(String propertyName, Collection<Object> values
+                  .map(v -> valueConverter.convert(v))
+                  .collect(Collectors.toList()));
+        }
+      } catch (IllegalArgumentException e) {
```
In case of multi value / repeated fields, an invalid value would result into skipping all values. Do we want to ignore just failing value or drop entire property?

cjohnson78 replied:
Yes, I see, it's probably better to skip only the bad value in the list. So, we probably cannot call valueConverter::convert directly. We would need to introduce a private method that does the conversion and swallows the exception. I will take another attempt at that.

cjohnson78 replied:
I have pushed this fix for multi value properties. Only the bad values are dropped, others are kept. If it reduces the list to empty, it returns null so the property is not set.

TanmayVartak reviewed (commented):
Thanks for your contribution. Can you also add some unit tests to validate this?

cjohnson78 committed:
Update indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/StructuredData.java

cjohnson78 committed:
Added logging when a conversion error is ignored.
Changed logic for repeating properties, so only bad values are rejected, the rest are kept.

cjohnson78 committed:
Merge remote-tracking branch 'origin/StructuredData-ignore-conversion-errors' into StructuredData-ignore-conversion-errors

# Conflicts:
#	indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/StructuredData.java

cjohnson78 committed:
Merge branch 'master' of https://github.com/google-cloudsearch/connector-sdk into StructuredData-ignore-conversion-errors

cjohnson78 committed:
Added test case.
I had to add setter for the configuration member because I could not call init(Schema)
and initFromConfiguration(indexingService) in the same test case.

cjohnson78 commented:
I have added the test case.
I had to add setter for the configuration member because I could not call init(Schema) 
and initFromConfiguration(indexingService) in the same test case.

donghanmiao commented on indexing/src/test/java/com/google/enterprise/cloudsearch/sdk/indexing/StructuredDataTest.java:
```diff
@@ -796,6 +760,28 @@ public void testDateTime_unparsedCharacters() throws IOException {
+    StructuredData.setIgnoreConversionErrors(true);
+    StructuredDataObject expected =
+            new StructuredDataObject()
+                    .setProperties(
```
the indentation looks different.  please use two spaces.
https://google.github.io/styleguide/javaguide.html#s4.2-block-indentation

cjohnson78 replied:
IntelliJ was having trouble auto-detecting 2-spaces. I have become aware of the issue and will correct the code.

cjohnson78 replied:
Fixed and pushed

donghanmiao commented on indexing/src/test/java/com/google/enterprise/cloudsearch/sdk/indexing/StructuredDataTest.java:
```diff
@@ -82,6 +36,16 @@
 
+import java.io.File;
+import java.io.IOException;
+import java.util.*;
```
same here, please refrain from using wildcard imports.  some IDE like intellij has such default setting but can be easily configured.

cjohnson78 replied:
Found that IntelliJ cannot disable the feature entirely, but you can set the threshold to 99 instances before wildcard is applied.

cjohnson78 replied:
fixed and pushed

cjohnson78 committed:
Fixed imports and whitespace differences

cjohnson78 committed:
Fixed imports and whitespace differences
